### PR TITLE
[ome] Invoke simplifyInstruction after lowering ownership and use replaceAllSimplifiedUsesAndErase instead of a manual RAUW.

### PR DESCRIPTION
**Pull request by gottesmm**
I am doing two things here:

1. simplifyInstruction expects its result to only be passed to
replaceAllSimplifiedUsesAndErase. This was an oversite.

2. I am upstreaming support for InstSimplify in OSSA (to finish sil-combine
work). The main assumption that inst-simplify is going to have is that it is
looking at SIL in correct ownership ssa or non-ownership ssa. It is not going to
be designed to work with an intermediate world that is the world where before
this PR it was being invoked (specifically, we have marked the function as not
being in OSSA but haven't finished transforming it into non-OSSA SIL). To avoid
this problem, I moved the simplification to the end of the pass after we have
done the initial traversal.

Should be NFCI.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
Not a big deal, but it's always helpful when renaming is done in a separate commit.

**Review by atrick — APPROVED** [review]
The change in this PR looks good. I asked for some clarification. The only issue I have is how badly SILBuilder is being abused for no reason. Why is the visitor passed a builder at all? The visitor should own a SILBuilderContext. It's trackingList should be that context's InsertedInsts. It can also hold on to the current builder, which can be instantiated in beforeVisit with the visitor's context.

Basically, don't call any setter functions on SILBuilder!

File: lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp [hunk 1]
```diff
@@ -45,180 +47,217 @@ DumpBefore("sil-dump-before-ome-to-path", llvm::cl::Hidden);
...
+  SILBuilder &builder;
+  SmallVector<SILInstruction *, 8> trackingList;
+  SmallBlotSetVector<SILInstruction *, 8> instructionsToSimplify;
+  SILOpenedArchetypesTracker openedArchetypesTracker;
+
+  OwnershipModelEliminatorVisitor(SILBuilder &newBuilder)
+      : builder(newBuilder),
+        openedArchetypesTracker(&newBuilder.getFunction()) {
+    newBuilder.setTrackingList(&trackingList);
+    newBuilder.setOpenedArchetypesTracker(&openedArchetypesTracker);
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:69]: I don't know why it was done this way TBH. I am fine changing it to however you want me to change it.

File: lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp [hunk 2]
```diff
@@ -45,180 +47,217 @@ DumpBefore("sil-dump-before-ome-to-path", llvm::cl::Hidden);
...
+      instructionsToSimplify.insert(trackingList.pop_back_val());
+    }
+  }
+
+  void beforeVisit(SILInstruction *instToVisit) {
+    // Add any elements to the tracking list that we currently have in the
+    // tracking list that we haven't added yet.
+    drainTrackingList();
+    builder.setInsertionPoint(instToVisit);
+    builder.setCurrentDebugScope(instToVisit->getDebugScope());
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:87]: This is old code. I will fix! Thanks for calling this out!

File: lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp [hunk 3]
```diff
@@ -245,158 +284,173 @@ bool OwnershipModelEliminatorVisitor::visitSwitchEnumInst(
...
+      SILInstruction *inst = &*ii;
+      ++ii;
+
+      madeChange |= visitor.visit(inst);
+    }
+  }
 
-      MadeChange |= Visitor.visit(I);
+  // Once we have finished processing all instructions, we should not be
+  // consistently in non-ossa form. Now go through any instructions and simplify
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:386]: @atrick when I was bringing up ossa part of my goal was zero diff (or as small as possible diff) after ownership lowering of our tests. With that in mind, I started to notice that a bunch of diff can from not running inst simplify's simplifications on split up destructures. It was completely a pragmatic decision.

That being said, OME is not an OSSA pass or a non-OSSA pass. It is more of a specialized transform that is in a between world while it is running. I don't think that is an assumption we want any of our utility code to have to deal with. Code should always be able to expect either fully non-ossa or ossa.

So it makes sense to hoist out this simplification code outside of the lowering so that it knows it structurally has non-ossa SIL. The actual thing we are simplifying are struct_extract from splitting destructures during the OME transform. So the simplification has to run after the transform on non-OSSA.

[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp]: I think I improved the wording.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
I am going to rebase this PR on top of #34844. #34844 fixes the bug that this PR was hitting (as one can see from the failing tests).

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci test linux platform

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
Woot all tests!

File: lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp [hunk 4]
```diff
@@ -57,16 +57,35 @@ namespace {
...
+  ///
+  /// Since this is syntactic and we assume that our caller is passing in a
+  /// lambda that if we inline will be eliminated, we mark this function always
+  /// inline.
+  template <typename ResultTy>
+  ResultTy LLVM_ATTRIBUTE_ALWAYS_INLINE
+  withBuilder(SILInstruction *insertPt,
+              llvm::function_ref<ResultTy(SILBuilder &, SILLocation)> visitor) {
+    SILBuilderWithScope builder(insertPt, builderCtx);
+    return visitor(builder, insertPt->getLoc());
```
Comment (atrick) [inline lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:88]: It's too bad this is so awkward. Ideally, building a single instruction should be able to naturally use function chaining:

`SILBuilder(inst, builderCtx).emitSomethingElse(operands)`

**Review by atrick — APPROVED** [review]
Thanks for fixing SILBuilder

File: lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp (hunk 4 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:88]: Yes. But today it is SILBuilderWithScope.

**Comment by gottesmm** [comment]
Going to rebase again.

**Comment by gottesmm** [comment]
Going to rebase on top of #34877

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
Rebased on top of main now that I landed the precursors.

**Comment by gottesmm** [comment]
@swift-ci test
